README.org
* jekyll-org

** Overview

This plugin adds [[http://orgmode.org/][Org mode]] support to [[http://jekyllrb.com][Jekyll]] and lets you write posts
and pages in Org.

** Related Projects 
   
What's the difference with org2jekll 
You don’t need to any org-publish-project-alist to starting write post, just start with any editor that org-mode supported

What’s the difference with org-jekyll?
You don’t need to add some alien yaml in your org-mode file. You add specific org-mode headers and this will be used to format the jekyll post.

What’s the difference with happyblogger?
No more external ruby script. 

** Requirements

- Jekyll
- org-ruby

** Installation
*** Short version

For experienced Jekyll users, you need to do 2 things to get
~jekyll-org~ to work :

- Include the gem in your ~_config.yml~
- Include the gem in Gemfile

*** Long version

To use ~jekyll-org~ with Jekyll, you need to have Ruby RubyGems and
Jekyll installed.  See how to do that [[http://jekyllrb.com/docs/installation/][here]].

Create a new Jekyll project ~my-site~ run:

#+begin_src sh
jekyll new my-site
#+end_src

Create a Gemfile in the root of the project, and add at least the
following lines:

#+begin_src ruby
source 'https://rubygems.org'

gem 'jekyll' , '>= 3.0.0'
gem 'jekyll-org', '>= 1.1.0'
#+end_src

Install the gems with bundler:

#+begin_src sh
bundle install
#+end_src

To use the new converter add the following line to your ~_config.yml~:

#+begin_src yml
plugins:
  - jekyll-org
#+end_src

:warning: If you are using Jekyll < 3.5.0 use the gems key instead of plugins.

** Usage

Create a new file with =.org= extension in =_posts=, and write the post
with Org. That is all! Generate your Jekyll site as you usually do.

** Front matter

Instead of YAML the front matter is configured in the usual Org way,
with no lines:

#+BEGIN_EXAMPLE
#+TITLE: Jekyll and Org together
#+LAYOUT: post
#+TAGS: jekyll org-mode "tag with spaces"

This is a blog post about Jekyll and Org mode.

** org-table demo 

   | a | b | c | d | e |
   |---+---+---+---+---|
   | 1 | 2 | 3 | 4 | 5 |

** Liquid demo 
   #+liquid: enabled
   #+foo: hello world
   {{ page.foo }}

   or

   {{ site.time | date_to_xmlschema }}
   
** Code highlighting
   Jekyll-org also offers powerful support for code snippets:
   #+begin_src  ruby 
     def print_hi(name)
       puts "Hi, #{name}"
     end
     print_hi('Tom')

     #=> prints 'Hi, Tom' to STDOUT.
   #+end_src
#+END_EXAMPLE

The value of =#+TAGS= is parsed into a list by splitting it on spaces,
tags containing spaces can be wrapped into quotes.

** Liquid templating

By default the all content is exported to raw HTML with org-ruby, but
you can add =#+liquid: whatevervalue== in the header.  Then you can use
[[http://docs.shopify.com/themes/liquid-documentation/basics][Liquid]] tags.

For example, if your Org file contains

#+BEGIN_EXAMPLE
#+liquid: enabled
#+foo: hello world

{{ page.foo }}

or

{{ site.time | date_to_xmlschema }}

#+END_EXAMPLE

then you will get output like

#+BEGIN_EXAMPLE
<p>hello world</p>
<p>or</p>
<p>2014-07-02T08:20:28+08:00</p>
#+END_EXAMPLE

*** Site wide config

Alternatively, if you'd rather enable liquid by default for every
org file being converted, you can enable it in your /_config.yml/
file.

#+BEGIN_EXAMPLE
org:
  liquid: true
#+END_EXAMPLE

** Source code highlighting

Add a source code block as you would in Org, for example Ruby:

#+BEGIN_EXAMPLE
#+BEGIN_SRC
 require 'rubygems'
 require 'org-ruby'
 data = IO.read(filename)
 puts Orgmode::Parser.new(data).to_html
#+END_SRC
#+END_EXAMPLE

Then the output will have code highlighting:

#+BEGIN_SRC ruby
 require 'rubygems'
 require 'org-ruby'
 data = IO.read(filename)
 puts Orgmode::Parser.new(data).to_html
#+END_SRC

** Author

eggcaker <eggcaker@gmail.com>

** License

MIT


History.md

1.1.1 / 2019-12-14
==================
  * add support for converting files in collections other than _posts

1.1.0 /2019-12-12 

  * add site wide configuration support through _config.yml


spec/helper.rb
require 'jekyll'
require File.expand_path('../lib/jekyll-org', File.dirname(__FILE__))

Jekyll.logger.log_level = :error

RSpec.configure do |config|
  config.run_all_when_everything_filtered = true
  config.filter_run :focus
  config.order = 'random'

  SOURCE_DIR = File.expand_path("../fixtures", __FILE__)
  DEST_DIR   = File.expand_path("../dest",     __FILE__)

  def source_dir(*files)
    File.join(SOURCE_DIR, *files)
  end

  def dest_dir(*files)
    File.join(DEST_DIR, *files)
  end
end


spec/test_jekyll-org.rb
require 'helper'

describe("should working with unit test") do
  it "added some test cases" do
  end
end


spec/.#test_jekyll-org-mode-converter.rb
eggcaker@Eggs-MacBook-Pro.local.472

convert.rb
require 'org-ruby'

if Jekyll::VERSION < "3.0"
  raise Jekyll::FatalException, "This version of jekyll-org is only compatible with Jekyll v3 and above."
end

module Jekyll
  class OrgConverter < Converter
    safe true

    priority :low

    def matches(ext)
      ext =~ /org/i
    end

    def output_ext(ext)
      ".html"
    end

    def convert(content)
      content
    end
  end

  module Filters
    def restify(input)
      site = @context.registers[:site]
      converter = site.find_converter_instance(Jekyll::OrgConverter)
      converter.convert(input)
    end
  end

  # This overrides having to use YAML in the posts.
  # Instead use settings from Org mode.
  class Document
    alias :_orig_read :read
    def read(opts = {})
      unless relative_path.end_with?(".org")
        return _orig_read(opts)
      end

      Jekyll.logger.debug "Reading:", relative_path

      self.content = File.read(path, Utils.merged_file_read_opts(site, opts))
      self.data ||= {}
      liquid_enabled = false

      org_text = Orgmode::Parser.new(content, { markup_file: "html.tags.yml" })
      org_text.in_buffer_settings.each_pair do |key, value|
        # Remove #+TITLE from the buffer settings to avoid double exporting
        org_text.in_buffer_settings.delete(key) if key =~ /title/i
        buffer_setting = key.downcase

        if buffer_setting == 'liquid'
          liquid_enabled = true
        end

        self.data[buffer_setting] = value
      end

      # Disable Liquid tags from the output by default or enabled with liquid_enabled tag

      if liquid_enabled
        self.content = org_text.to_html
        self.content = self.content.gsub("&#8216;","'")
        self.content = self.content.gsub("&#8217;", "'")
      else
        self.content = <<ORG
{% raw %}
#{org_text.to_html}
{% endraw %}
ORG
      end

      post_read
    rescue => e
      puts "Error converting file #{relative_path}: #{e.message} #{e.backtrace}"
    end
  end
end


lib/jekyll-org.rb
if Jekyll::VERSION < '3.0'
  raise Jekyll::FatalException, 'This version of jekyll-org is only compatible with Jekyll v3 and above.'
end

require_relative './jekyll-org/converter'

module Jekyll
  module Filters
    def orgify(input)
      site      = @context.registers[:site]
      converter = site.find_converter_instance(Jekyll::Converters::Org)
      converter.convert(input)
    end
  end

  module OrgDocument
    def org_file?
      Jekyll::Converters::Org.matches(extname)
    end

    ## override: read file & parse YAML... in this case, don't parse YAML
    # see also: https://github.com/jekyll/jekyll/blob/master/lib/jekyll/document.rb
    def read_content(opts = {})
      return super unless org_file? # defer to default behaviour when not org file

      self.content = File.read(path, Utils.merged_file_read_opts(site, opts))
      converter = site.find_converter_instance(Jekyll::Converters::Org)
      parser, settings = converter.parse_org(self.content)

      @data = Utils.deep_merge_hashes(self.data, settings)
      self.content = converter.actually_convert(parser)
    end
  end

  module OrgPage
    # Don't parse YAML blocks when working with an org file, parse it as
    # an org file and then extracts the settings from there.
    def read_yaml(base, name, opts = {})
      extname  = File.extname(name)
      return super unless Jekyll::Converters::Org.matches(extname)
      filename = File.join(base, name)

      self.data ||= Hash.new()
      self.content = File.read(@path || site.in_source_dir(base, name),
                               Utils.merged_file_read_opts(site, opts))
      converter = site.find_converter_instance(Jekyll::Converters::Org)
      parser, settings = converter.parse_org(self.content)

      self.content = converter.actually_convert(parser)
      self.data = Utils.deep_merge_hashes(self.data, settings)
    end
  end

  class Collection
    # make jekyll collections aside from _posts also convert org files.
    def read
      filtered_entries.each do |file_path|
        full_path = collection_dir(file_path)
        next if File.directory?(full_path)

        if Utils.has_yaml_header?(full_path) ||
           Jekyll::Converters::Org.matches_path(full_path)
          read_document(full_path)
        else
          read_static_file(file_path, full_path)
        end
      end

      # TODO remove support for jekyll 3 on 4s release
      (Jekyll::VERSION < '4.0') ? docs.sort! : sort_docs!
    end
  end

  class Reader
    def retrieve_org_files(dir, files)
      retrieve_pages(dir, files)
    end

    # don't interpret org files without YAML as static files.
    def retrieve_static_files(dir, files)
      org_files, static_files = files.partition(&Jekyll::Converters::Org.method(:matches_path))

      retrieve_org_files(dir, org_files) unless org_files.empty?
      site.static_files.concat(StaticFileReader.new(site, dir).read(static_files))
    end
  end
end

Jekyll::Document.prepend(Jekyll::OrgDocument)
Jekyll::Page.prepend(Jekyll::OrgPage)


lib/jekyll-org/converter.rb
require 'csv'
require 'org-ruby'
require_relative './utils'

module Jekyll
  module Converters
    class Org < Converter
      include Jekyll::OrgUtils

      safe     true
      priority :low

      # true if current file is an org file
      def self.matches(ext)
        ext =~ /org/i
      end

      # matches, but accepts complete path
      def self.matches_path(path)
        matches(File.extname(path))
      end

      def matches(ext)
        self.class.matches ext
      end

      def output_ext(ext)
        '.html'
      end

      # because by the time an org file reaches the
      # converter... it will have already been converted.
      def convert(content)
        content
      end

      def actually_convert(content)
        case content
        when Orgmode::Parser
          if liquid_enabled?(content)
            content.to_html.gsub("&#8216;", "'").gsub("&#8217;", "'")
          else
            '{% raw %} ' + content.to_html + ' {% endraw %}'
          end
        else
          parser, variables = parse_org(content)
          convert(parser)
        end
      end

      def parse_org(content)
        parser   = Orgmode::Parser.new(content, parser_options)
        settings = { :liquid => org_config.fetch("liquid", false) }

        parser.in_buffer_settings.each_pair do |key, value|
          # Remove #+TITLE from the buffer settings to avoid double exporting
          parser.in_buffer_settings.delete(key) if key =~ /title/i
          assign_setting(key, value, settings)
        end

        return parser, settings
      end

      private

      def config_liquid_enabled?
        org_config.fetch("liquid", false)
      end

      def liquid_enabled?(parser)
        tuple = parser.in_buffer_settings.each_pair.find do |key, _|
          key =~ /liquid/i
        end

        if tuple.nil?
            config_liquid_enabled? # default value, as per config
        else
            _parse_boolean(tuple[1], "unknown LIQUID setting")
        end
      end

      # see https://github.com/bdewey/org-ruby/blob/master/lib/org-ruby/parser.rb
      def parser_options
        org_config.fetch("parser_options", { markup_file: 'html.tags.yml' })
      end

      def assign_setting(key, value, settings)
        key = key.downcase

        case key
        when "liquid"
          value = _parse_boolean(value, "unknown LIQUID setting")
          settings[:liquid] = value unless value.nil?
        when "tags", "categories"
          # Parse a string of tags separated by spaces into a list.
          # Tags containing spaces can be wrapped in quotes,
          # e.g. '#+TAGS: foo "with spaces"'.
          #
          # The easiest way to do this is to use rubys builtin csv parser
          # and use spaces instead of commas as column separator.
          settings[key] = CSV::parse_line(value, col_sep: ' ')
        else
          value_bool = _parse_boolean(value)
          settings[key] = (value_bool.nil?) ? value : value_bool
        end
      end

      @@truthy_regexps = [/enabled/,  /yes/, /true/]
      @@falsy_regexps  = [/disabled/, /no/,  /false/]

      def _parse_boolean(value, error_msg=nil)
        case value.downcase
        when *@@truthy_regexps
          true
        when *@@falsy_regexps
          false
        else
          unless error_msg.nil?
            Jekyll.logger.warn("OrgDocument:", error_msg + ": #{value}")
          end
        end
      end
    end
  end
end
